PMM-5757 diagnostic data collector

percona-csalguero opened the pull request:
https://jira.percona.com/browse/PMM-5757

- Diagnostic data collector
- Replicaset status collector
- Collstats collector
- Server status collector
- Main parameters
- HTTP server that will support having multiple paths to expose metrics

percona-csalguero committed:
Sandbox, Connection and tests

percona-csalguero committed:
WIP Test push pemissions

percona-csalguero committed:
.goreleaser, version and debug

percona-csalguero committed:
Metrics renaming & collecting

percona-csalguero committed:
PMM-5739 Metrics renamer

percona-csalguero committed:
PMM-5757 diagnostic data collector

- Diagnostic data collector
- Replicaset status collector
- Collstats collector
- Server status collector
- Main parameters
- HTTP server that will support having multiple paths to expose metrics

percona-csalguero committed:
Fixed conflicts

percona-csalguero committed:
Updated job definition

percona-csalguero committed:
Updated test

percona-csalguero committed:
Merge branch 'PMM-5739' into PMM-5757_getDiagnosticData

percona-csalguero committed:
PMM-5757 Using shared http server

AlekSi commented on exporter/exporter.go:
```diff
@@ -13,36 +18,74 @@ const (
+	return exp, nil
+}
+
+func (e *Exporter) Run() {
```
That really should be in #4

BupycHuk reviewed (changes requested):
Please fix merge conflicts

cfsalguero committed:
Merge branch 'master' into merge-test

cfsalguero committed:
PMM-5757 Merged and fixed all tests

cfsalguero committed:
PMM-5757 Fixed tests & code clean up

cfsalguero committed:
PMM-5757 Linter fixes

cfsalguero committed:
PMM-5757 Added to gitignore

BupycHuk commented on exporter/collstats_collector.go:
```diff
@@ -0,0 +1,41 @@
+		cmd := bson.D{{Key: "collStats", Value: parts[1]}}
+		res := d.client.Database(parts[0]).RunCommand(d.ctx, cmd)
+		var m bson.M
+		if err := res.Decode(&m); err != nil {
```
maybe it's better to disable this linter

AlekSi replied:
Agree

@percona-csalguero ^

AlekSi replied:
@percona-csalguero ^^^

BupycHuk commented on exporter/replset_status_collector_test.go:
```diff
@@ -0,0 +1,50 @@
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	client := getTestClient(ctx, t)
```
Can we add test for ReplsetStatusCollector for cases with no replicaset?

BupycHuk commented on exporter/serverstatus_collector_test.go:
```diff
@@ -0,0 +1,51 @@
+
+	client := getTestClient(ctx, t)
+
+	c := &diagnosticDataCollector{
```
`diagnosticDataCollector`?

BupycHuk reviewed (changes requested):
I remember that you said that you will add API as part of this PR, but I don't see it. Will you make separate PR?

cfsalguero committed:
PMM-5757 Fixes for CR

- New package to group testing util functions.
- Fixed ServerStatus collector
- New test for replicaset collector while connected to a non sharded
server

BupycHuk commented on internal/tu/testutils.go:
```diff
@@ -0,0 +1,63 @@
+package tu
```
what does tu means?

percona-csalguero replied:
test utils

BupycHuk commented on internal/tu/testutils.go:
```diff
@@ -0,0 +1,63 @@
+package tu
```
please fix golangci linter comments or add ignore comments.

BupycHuk commented on exporter/replset_status_collector_test.go:
```diff
@@ -0,0 +1,68 @@
+
+	expected := strings.NewReader(``)
+	err := testutil.CollectAndCompare(c, expected)
+	assert.Error(t, err)
```
let's use `assert.EqualError`

percona-csalguero replied:
Improved to detect the Replicaset not yet initialized error

BupycHuk reviewed (changes requested):
I don't see anything related to `HTTP server that will support having multiple paths to expose metrics`

cfsalguero committed:
PMM-5757 Fixes for linters

cfsalguero committed:
PMM-5757 Improved replicaset collector

percona-csalguero committed:
Merge branch 'master' into PMM-5757_getDiagnosticData

BupycHuk commented:
please fix a build.

AlekSi reviewed (changes requested):
CI should pass

cfsalguero committed:
PMM-5757 Updated deps

percona-csalguero commented:
@AlekSi @BupycHuk , CI checks passing

cfsalguero committed:
PMM-5757 Disabled WSL linter

cfsalguero committed:
PMM-5757 Added license headers

cfsalguero committed:
PMM-5757 Updated license

askomorokhov reviewed (approved):
👍
